File: app/review.py
"""Revisao de partida: avalia cada lance com Stockfish e classifica no estilo
do "Game Review" do chess.com.

A base de calculo (perda por lance a partir da avaliacao antes/depois, deteccao
de fase, livro de aberturas em EPD) vem da skill hhkarimi/claude-chess-skills;
a camada de win%, precisao e rotulos (Brilhante, Otimo, Melhor, ...) segue as
formulas publicas usadas por lichess/chess.com.
"""

from __future__ import annotations

import math
from pathlib import Path

import chess
import chess.engine
import chess.pgn

from . import chesscom

ASSETS = Path(__file__).resolve().parent / "assets"

# Avaliacao maxima considerada: acima disso a partida ja esta decidida e a
# diferenca nao deve poluir as medias.
EVAL_CAP = 1000
MATE_SCORE = 10000

PIECE_VALUE = {
    chess.PAWN: 1,
    chess.KNIGHT: 3,
    chess.BISHOP: 3,
    chess.ROOK: 5,
    chess.QUEEN: 9,
    chess.KING: 0,
}

# Perda de win% (0-100) que define cada rotulo.
THRESHOLDS = [
    (2.0, "excellent"),
    (5.0, "good"),
    (10.0, "inaccuracy"),
    (20.0, "mistake"),
]

LABELS_PT = {
    "brilliant": "Brilhante",
    "great": "Otimo",
    "best": "Melhor",
    "excellent": "Excelente",
    "good": "Bom",
    "book": "Livro",
    "inaccuracy": "Impreciso",
    "mistake": "Erro",
    "miss": "Chance perdida",
    "blunder": "Erro grave",
}


# ---------------------------------------------------------------------------
# Conversoes de avaliacao
# ---------------------------------------------------------------------------

def cp_white(score: chess.engine.PovScore) -> int:
    """Avaliacao em centipeoes do ponto de vista das brancas, com teto."""
    wcp = score.white().score(mate_score=MATE_SCORE)
    if wcp is None:
        return 0
    return max(-EVAL_CAP, min(EVAL_CAP, wcp))


def win_pct(cp: int) -> float:
    """Chance de vitoria (0-100) das brancas para uma avaliacao em centipeoes."""
    cp = max(-EVAL_CAP, min(EVAL_CAP, cp))
    return 50 + 50 * (2 / (1 + math.exp(-0.00368208 * cp)) - 1)


def win_pct_for(cp_w: int, white_to_move: bool) -> float:
    """Chance de vitoria do lado que esta jogando."""
    w = win_pct(cp_w)
    return w if white_to_move else 100 - w


def move_accuracy(win_before: float, win_after: float) -> float:
    """Precisao (0-100) de um lance, pela curva usada no lichess."""
    delta = max(0.0, win_before - win_after)
    acc = 103.1668 * math.exp(-0.04354 * delta) - 3.1669
    return max(0.0, min(100.0, acc))


def game_accuracy(accuracies: list[float], win_series: list[float]) -> float | None:
    """Media ponderada por volatilidade + media harmonica (metodo do lichess)."""
    if not accuracies:
        return None
    window = max(2, min(8, len(win_series) // 10))
    weights = []
    for i in range(len(accuracies)):
        lo = max(0, i - window)
        hi = min(len(win_series), i + window + 1)
        chunk = win_series[lo:hi] or [50.0]
        mean = sum(chunk) / len(chunk)
        var = sum((x - mean) ** 2 for x in chunk) / len(chunk)
        weights.append(max(0.5, min(12.0, math.sqrt(var))))

    total_w = sum(weights) or 1.0
    weighted = sum(a * w for a, w in zip(accuracies, weights)) / total_w
    harmonic = len(accuracies) / sum(1 / max(a, 1e-6) for a in accuracies)
    return round(max(0.0, min(100.0, (weighted + harmonic) / 2)), 1)


# ---------------------------------------------------------------------------
# Heuristicas de posicao
# ---------------------------------------------------------------------------

def load_book() -> set:
    path = ASSETS / "openings_book.txt"
    if not path.exists():
        return set()
    return {
        line.strip()
        for line in path.read_text(encoding="utf-8").splitlines()
        if line.strip() and not line.startswith("#")
    }


def phase_of(board: chess.Board) -> str:
    """Abertura / meio-jogo / final por material sem peoes e sem reis."""
    material = sum(
        PIECE_VALUE[p.piece_type]
        for p in board.piece_map().values()
        if p.piece_type not in (chess.PAWN, chess.KING)
    )
    if board.fullmove_number <= 12 and material >= 26:
        return "opening"
    if material <= 12:
        return "endgame"
    return "middlegame"


def material_balance(board: chess.Board, color: chess.Color) -> int:
    mine = sum(
        PIECE_VALUE[p.piece_type] for p in board.piece_map().values() if p.color == color
    )
    theirs = sum(
        PIECE_VALUE[p.piece_type] for p in board.piece_map().values() if p.color != color
    )
    return mine - theirs


def _is_sacrifice(board_before: chess.Board, move: chess.Move, pv: list) -> bool:
    """Sacrificio: depois do lance e das melhores respostas, quem jogou fica com
    pelo menos um leve a menos de material."""
    mover = board_before.turn
    before = material_balance(board_before, mover)

    probe = board_before.copy(stack=False)
    probe.push(move)
    plies = 0
    for pv_move in pv:
        if plies >= 3 or pv_move not in probe.legal_moves:
            break
        probe.push(pv_move)
        plies += 1
    return material_balance(probe, mover) - before <= -2


# ---------------------------------------------------------------------------
# Classificacao de lances
# ---------------------------------------------------------------------------

def classify(
    *,
    board_before: chess.Board,
    move: chess.Move,
    best_move,
    win_before: float,
    win_after: float,
    second_best_win,
    in_book: bool,
    mate_before,
    mate_after,
    pv_after: list,
) -> str:
    loss = max(0.0, win_before - win_after)
    is_best = best_move is not None and move == best_move

    if in_book:
        return "book"

    if is_best and loss <= 2.0:
        if win_after >= 45 and _is_sacrifice(board_before, move, pv_after):
            return "brilliant"
        # Unico lance que segura a posicao: a segunda opcao e muito pior, e a
        # partida ainda estava em aberto (nao vale em posicao ja decidida).
        if (
            second_best_win is not None
            and (win_after - second_best_win) >= 20
            and 15 <= win_after <= 92
        ):
            return "great"
        return "best"

    # Tinha mate forcado e deixou escapar.
    if mate_before is not None and mate_before > 0 and mate_after is None and loss >= 5:
        return "miss"

    for limit, label in THRESHOLDS:
        if loss < limit:
            return label
    return "blunder"


# ---------------------------------------------------------------------------
# Analise da partida
# ---------------------------------------------------------------------------

def analyze_game(pgn: str, my_color: str, depth: int = 14, engine=None):
    """Gera eventos de progresso e, no fim, a revisao completa.

    Cada yield e um dict: {"type": "progress"|"done"|"error", ...}
    """
    from .engine import open_engine

    game = chesscom.parse_pgn(pgn)
    book = load_book()

    nodes = list(game.mainline())
    total = len(nodes)
    if total == 0:
        yield {"type": "error", "message": "Partida sem lances."}
        return

    own_engine = engine is None
    engine = engine or open_engine()
    limit = chess.engine.Limit(depth=depth)

    try:
        board = game.board()
        evals = []  # avaliacao de cada posicao, incluindo a inicial

        def evaluate(b: chess.Board) -> dict:
            if b.is_game_over(claim_draw=True):
                outcome = b.outcome(claim_draw=True)
                if outcome and outcome.winner is not None:
                    cp = EVAL_CAP if outcome.winner == chess.WHITE else -EVAL_CAP
                else:
                    cp = 0
                return {"cp_white": cp, "best": None, "second_win": None, "pv": [], "mate": None}

            infos = engine.analyse(b, limit, multipv=2)
            first = infos[0]
            cp_w = cp_white(first["score"])
            pv = list(first.get("pv", []))
            best = pv[0] if pv else None
            mate = first["score"].pov(b.turn).mate()

            second_win = None
            if len(infos) > 1:
                second_win = win_pct_for(cp_white(infos[1]["score"]), b.turn)

            return {
                "cp_white": cp_w,
                "best": best,
                "second_win": second_win,
                "pv": pv,
                "mate": mate,
            }

        evals.append(evaluate(board))
        yield {"type": "progress", "done": 0, "total": total}

        moves = []
        for idx, node in enumerate(nodes):
            move = node.move
            board_before = board.copy(stack=False)
            san = board_before.san(move)
            board.push(move)

            after = evaluate(board)
            evals.append(after)

            before = evals[idx]
            mover_is_white = board_before.turn == chess.WHITE
            wb = win_pct_for(before["cp_white"], mover_is_white)
            wa = win_pct_for(after["cp_white"], mover_is_white)

            in_book = idx < 24 and board.epd() in book
            mate_after_mover = -after["mate"] if after["mate"] is not None else None
            label = classify(
                board_before=board_before,
                move=move,
                best_move=before["best"],
                win_before=wb,
                win_after=wa,
                second_best_win=before["second_win"],
                in_book=in_book,
                mate_before=before["mate"],
                mate_after=mate_after_mover,
                pv_after=after["pv"],
            )

            # Como o adversario pune o lance: a linha principal a partir da
            # posicao resultante. Alimenta a deteccao de padroes de erro.
            # 6 meios-lances: com 4 o material perdido em sequencias um pouco
            # mais longas passava batido e caia no balde generico.
            punish_line = []
            probe = board.copy(stack=False)
            for pv_move in after["pv"][:6]:
                if pv_move not in probe.legal_moves:
                    break
                punish_line.append(probe.san(pv_move))
                probe.push(pv_move)
            material_swing = material_balance(probe, board_before.turn) - material_balance(
                board_before, board_before.turn
            )
            punish_is_check = bool(after["pv"]) and board.gives_check(after["pv"][0])
            punish_is_capture = bool(after["pv"]) and board.is_capture(after["pv"][0])

            best_san = None
            best_line = []
            if before["best"] is not None and before["best"] != move:
                best_san = board_before.san(before["best"])
                probe = board_before.copy(stack=False)
                for pv_move in before["pv"][:6]:
                    if pv_move not in probe.legal_moves:
                        break
                    best_line.append(probe.san(pv_move))
                    probe.push(pv_move)

            moves.append(
                {
                    "ply": idx + 1,
                    "move_no": board_before.fullmove_number,
                    "side": "white" if mover_is_white else "black",
                    "san": san,
                    "uci": move.uci(),
                    "from": chess.square_name(move.from_square),
                    "to": chess.square_name(move.to_square),
                    "fen_before": board_before.fen(),
                    "fen_after": board.fen(),
                    "eval_before": before["cp_white"],
                    "eval_after": after["cp_white"],
                    "mate_after": mate_after_mover,
                    "win_before": round(wb, 1),
                    "win_after": round(wa, 1),
                    "cpl": round(max(0.0, wb - wa) * 10),
                    "accuracy": round(move_accuracy(wb, wa), 1),
                    "class": label,
                    "label": LABELS_PT[label],
                    "best_move": best_san,
                    "best_uci": before["best"].uci() if before["best"] else None,
                    "best_line": best_line,
                    "punish_line": punish_line,
                    "punish_is_check": punish_is_check,
                    "punish_is_capture": punish_is_capture,
                    "material_swing": material_swing,
                    "phase": phase_of(board_before),
                    "clock": chesscom.clock_seconds(node.comment),
                    "is_capture": board_before.is_capture(move),
                    "is_check": board.is_check(),
                }
            )

            yield {"type": "progress", "done": idx + 1, "total": total}

        yield {"type": "done", "review": summarize(game, moves, my_color, depth)}
    finally:
        if own_engine:
            engine.quit()


def summarize(game: chess.pgn.Game, moves: list, my_color: str, depth: int) -> dict:
    """Monta os agregados que a tela de revisao consome."""
    per_side = {}
    for side in ("white", "black"):
        side_moves = [m for m in moves if m["side"] == side]
        counts = {k: 0 for k in LABELS_PT}
        for m in side_moves:
            counts[m["class"]] += 1

        wins = [
            m["win_before"] if m["side"] == side else 100 - m["win_before"] for m in moves
        ]
        per_side[side] = {
            "counts": counts,
            "accuracy": game_accuracy([m["accuracy"] for m in side_moves], wins),
            "avg_cpl": (
                round(sum(m["cpl"] for m in side_moves) / len(side_moves))
                if side_moves
                else None
            ),
            "avg_cpl_by_phase": _avg_by_phase(side_moves),
        }

    key_moments = sorted(
        (m for m in moves if m["class"] in ("blunder", "mistake", "miss")),
        key=lambda m: m["win_before"] - m["win_after"],
        reverse=True,
    )[:6]

    headers = game.headers
    return {
        "white": headers.get("White", "Brancas"),
        "black": headers.get("Black", "Pretas"),
        "white_elo": headers.get("WhiteElo"),
        "black_elo": headers.get("BlackElo"),
        "result": headers.get("Result", "*"),
        "opening": chesscom.opening_name(dict(headers)),
        "date": headers.get("UTCDate") or headers.get("Date", ""),
        "my_color": my_color,
        "depth": depth,
        "moves": moves,
        "summary": per_side,
        "key_moments": key_moments,
        "eval_series": [
            {"ply": 0, "eval": moves[0]["eval_before"] if moves else 0},
            *[{"ply": m["ply"], "eval": m["eval_after"]} for m in moves],
        ],
    }


def _avg_by_phase(side_moves: list) -> dict:
    out = {}
    for phase in ("opening", "middlegame", "endgame"):
        vals = [m["cpl"] for m in side_moves if m["phase"] == phase]
        if vals:
            out[phase] = round(sum(vals) / len(vals))
    return out

File: web/review.js
/* Revisao de partida no navegador. Porte de app/review.py.
 *
 * Mesmas formulas: avaliacao -> chance de vitoria (curva do lichess), precisao
 * por lance, e os rotulos no estilo do Game Review do chess.com.
 */

import { Chess } from "./vendor/chess.js";
import { clockSeconds, openingName } from "./chesscom.js";

const EVAL_CAP = 1000;

const PIECE_VALUE = { p: 1, n: 3, b: 3, r: 5, q: 9, k: 0 };

// Perda de win% (0-100) que define cada rotulo.
const THRESHOLDS = [
  [2.0, "excellent"],
  [5.0, "good"],
  [10.0, "inaccuracy"],
  [20.0, "mistake"],
];

export const LABELS_PT = {
  brilliant: "Brilhante",
  great: "Otimo",
  best: "Melhor",
  excellent: "Excelente",
  good: "Bom",
  book: "Livro",
  inaccuracy: "Impreciso",
  mistake: "Erro",
  miss: "Chance perdida",
  blunder: "Erro grave",
};

/* ------------------------------------------------------------ avaliacao */

/** Converte o score do UCI (POV de quem joga) para centipeoes POV das brancas. */
export function toWhiteCp({ cp, mate }, whiteToMove) {
  let value;
  if (mate !== null && mate !== undefined) value = mate > 0 ? EVAL_CAP : -EVAL_CAP;
  else if (cp === null || cp === undefined) value = 0;
  else value = cp;
  const white = whiteToMove ? value : -value;
  return Math.max(-EVAL_CAP, Math.min(EVAL_CAP, white));
}

export function winPct(cp) {
  const c = Math.max(-EVAL_CAP, Math.min(EVAL_CAP, cp));
  return 50 + 50 * (2 / (1 + Math.exp(-0.00368208 * c)) - 1);
}

export function winPctFor(cpWhite, whiteToMove) {
  const w = winPct(cpWhite);
  return whiteToMove ? w : 100 - w;
}

export function moveAccuracy(winBefore, winAfter) {
  const delta = Math.max(0, winBefore - winAfter);
  const acc = 103.1668 * Math.exp(-0.04354 * delta) - 3.1669;
  return Math.max(0, Math.min(100, acc));
}

export function gameAccuracy(accuracies, winSeries) {
  if (!accuracies.length) return null;
  const window = Math.max(2, Math.min(8, Math.floor(winSeries.length / 10)));

  const weights = accuracies.map((_, i) => {
    const chunk = winSeries.slice(Math.max(0, i - window), i + window + 1);
    if (!chunk.length) return 0.5;
    const mean = chunk.reduce((a, b) => a + b, 0) / chunk.length;
    const varr = chunk.reduce((a, b) => a + (b - mean) ** 2, 0) / chunk.length;
    return Math.max(0.5, Math.min(12, Math.sqrt(varr)));
  });

  const totalW = weights.reduce((a, b) => a + b, 0) || 1;
  const weighted = accuracies.reduce((a, acc, i) => a + acc * weights[i], 0) / totalW;
  const harmonic = accuracies.length / accuracies.reduce((a, x) => a + 1 / Math.max(x, 1e-6), 0);
  return Math.round(Math.max(0, Math.min(100, (weighted + harmonic) / 2)) * 10) / 10;
}

/* --------------------------------------------------------- heuristicas */

let bookCache = null;

export async function loadBook() {
  if (bookCache) return bookCache;
  try {
    const text = await fetch("openings_book.txt").then((r) => r.text());
    bookCache = new Set(
      text.split("\n").map((l) => l.trim()).filter((l) => l && !l.startsWith("#"))
    );
  } catch {
    bookCache = new Set();
  }
  return bookCache;
}

/** EPD = os 4 primeiros campos do FEN (sem os contadores de lance). */
function epd(fen) {
  return fen.split(" ").slice(0, 4).join(" ");
}

function boardCounts(fen) {
  const rows = fen.split(" ")[0];
  const out = [];
  for (const ch of rows) {
    if (/[a-zA-Z]/.test(ch)) out.push(ch);
  }
  return out;
}

export function phaseOf(fen) {
  const fullmove = parseInt(fen.split(" ")[5], 10) || 1;
  const material = boardCounts(fen)
    .filter((c) => !"pPkK".includes(c))
    .reduce((a, c) => a + PIECE_VALUE[c.toLowerCase()], 0);

  if (fullmove <= 12 && material >= 26) return "opening";
  if (material <= 12) return "endgame";
  return "middlegame";
}

/** Saldo de material do ponto de vista de `color` ('w' | 'b'). */
function materialBalance(fen, color) {
  let mine = 0;
  let theirs = 0;
  for (const c of boardCounts(fen)) {
    const value = PIECE_VALUE[c.toLowerCase()];
    const isWhite = c === c.toUpperCase();
    if ((color === "w") === isWhite) mine += value;
    else theirs += value;
  }
  return mine - theirs;
}

function pushUci(chess, uci) {
  try {
    return chess.move({
      from: uci.slice(0, 2),
      to: uci.slice(2, 4),
      promotion: uci.length > 4 ? uci[4] : undefined,
    });
  } catch {
    return null;
  }
}

/** Aplica ate `max` lances da linha e devolve os SAN e o FEN final. */
function walkLine(fromFen, uciMoves, max) {
  const chess = new Chess(fromFen);
  const san = [];
  for (const uci of uciMoves.slice(0, max)) {
    const mv = pushUci(chess, uci);
    if (!mv) break;
    san.push(mv.san);
  }
  return { san, fen: chess.fen() };
}

function isSacrifice(fenBefore, moveUci, pvAfter) {
  const chess = new Chess(fenBefore);
  const mover = chess.turn();
  const before = materialBalance(fenBefore, mover);
  if (!pushUci(chess, moveUci)) return false;

  const { fen } = walkLine(chess.fen(), pvAfter, 3);
  return materialBalance(fen, mover) - before <= -2;
}

/* -------------------------------------------------------- classificacao */

function classify(o) {
  const loss = Math.max(0, o.winBefore - o.winAfter);
  const isBest = o.bestUci && o.moveUci === o.bestUci;

  if (o.inBook) return "book";

  if (isBest && loss <= 2.0) {
    if (o.winAfter >= 45 && isSacrifice(o.fenBefore, o.moveUci, o.pvAfter)) return "brilliant";
    if (o.secondBestWin !== null && o.winAfter - o.secondBestWin >= 20 &&
        o.winAfter >= 15 && o.winAfter <= 92) {
      return "great";
    }
    return "best";
  }

  if (o.mateBefore !== null && o.mateBefore > 0 && o.mateAfter === null && loss >= 5) {
    return "miss";
  }

  for (const [limit, label] of THRESHOLDS) {
    if (loss < limit) return label;
  }
  return "blunder";
}

/* -------------------------------------------------------------- analise */

/**
 * Analisa a partida inteira.
 * @param onProgress ({done, total}) chamado a cada lance.
 */
export async function analyzeGame(engine, pgn, myColor, { depth = 12, onProgress, signal } = {}) {
  const book = await loadBook();

  const game = new Chess();
  game.loadPgn(pgn);
  const history = game.history({ verbose: true });
  if (!history.length) throw new Error("Partida sem lances.");

  // getComments() indexa pelo FEN resultante do lance - e ali que vem o relogio.
  const clocks = new Map();
  for (const { fen, comment } of game.getComments()) clocks.set(fen, clockSeconds(comment));

  const evaluate = async (fen) => {
    const probe = new Chess(fen);
    if (probe.isGameOver()) {
      let cp = 0;
      if (probe.isCheckmate()) cp = probe.turn() === "w" ? -EVAL_CAP : EVAL_CAP;
      return { cpWhite: cp, bestUci: null, secondWin: null, pv: [], mate: null };
    }

    const infos = await engine.analyse(fen, { depth, multipv: 2 });
    const whiteToMove = probe.turn() === "w";
    const first = infos[0] || { cp: 0, mate: null, pv: [] };

    return {
      cpWhite: toWhiteCp(first, whiteToMove),
      bestUci: first.pv[0] || null,
      secondWin: infos[1]
        ? winPctFor(toWhiteCp(infos[1], whiteToMove), whiteToMove)
        : null,
      pv: first.pv,
      mate: first.mate ?? null,
    };
  };

  const evals = [await evaluate(history[0].before)];
  onProgress?.({ done: 0, total: history.length });

  const moves = [];
  for (let idx = 0; idx < history.length; idx++) {
    if (signal?.aborted) throw new DOMException("cancelado", "AbortError");

    const h = history[idx];
    const after = await evaluate(h.after);
    evals.push(after);

    const before = evals[idx];
    const moverIsWhite = h.color === "w";
    const wb = winPctFor(before.cpWhite, moverIsWhite);
    const wa = winPctFor(after.cpWhite, moverIsWhite);

    const inBook = idx < 24 && book.has(epd(h.after));
    const mateAfterMover = after.mate === null ? null : -after.mate;

    const label = classify({
      fenBefore: h.before,
      moveUci: h.lan,
      bestUci: before.bestUci,
      winBefore: wb,
      winAfter: wa,
      secondBestWin: before.secondWin,
      inBook,
      mateBefore: before.mate,
      mateAfter: mateAfterMover,
      pvAfter: after.pv,
    });

    // Como o adversario pune: 6 meios-lances a partir da posicao resultante.
    const punish = walkLine(h.after, after.pv, 6);
    const materialSwing =
      materialBalance(punish.fen, h.color) - materialBalance(h.before, h.color);

    let punishIsCheck = false;
    let punishIsCapture = false;
    if (after.pv.length) {
      const probe = new Chess(h.after);
      const mv = pushUci(probe, after.pv[0]);
      if (mv) {
        punishIsCheck = probe.isCheck();
        punishIsCapture = mv.flags.includes("c") || mv.flags.includes("e");
      }
    }

    let bestSan = null;
    let bestLine = [];
    if (before.bestUci && before.bestUci !== h.lan) {
      const line = walkLine(h.before, before.pv, 6);
      bestSan = line.san[0] || null;
      bestLine = line.san;
    }

    moves.push({
      ply: idx + 1,
      move_no: Number(h.before.split(" ")[5]) || 1,
      side: moverIsWhite ? "white" : "black",
      san: h.san,
      uci: h.lan,
      from: h.from,
      to: h.to,
      fen_before: h.before,
      fen_after: h.after,
      eval_before: before.cpWhite,
      eval_after: after.cpWhite,
      mate_after: mateAfterMover,
      win_before: Math.round(wb * 10) / 10,
      win_after: Math.round(wa * 10) / 10,
      cpl: Math.round(Math.max(0, wb - wa) * 10),
      accuracy: Math.round(moveAccuracy(wb, wa) * 10) / 10,
      class: label,
      label: LABELS_PT[label],
      best_move: bestSan,
      best_uci: before.bestUci,
      best_line: bestLine,
      punish_line: punish.san,
      punish_is_check: punishIsCheck,
      punish_is_capture: punishIsCapture,
      material_swing: materialSwing,
      phase: phaseOf(h.before),
      clock: clocks.get(h.after) ?? null,
      is_capture: h.flags.includes("c") || h.flags.includes("e"),
      is_check: new Chess(h.after).isCheck(),
    });

    onProgress?.({ done: idx + 1, total: history.length });
  }

  return summarize(game, moves, myColor, depth);
}

function summarize(game, moves, myColor, depth) {
  const summary = {};
  for (const side of ["white", "black"]) {
    const sideMoves = moves.filter((m) => m.side === side);
    const counts = Object.fromEntries(Object.keys(LABELS_PT).map((k) => [k, 0]));
    for (const m of sideMoves) counts[m.class]++;

    const wins = moves.map((m) =>
      m.side === side ? m.win_before : 100 - m.win_before
    );

    summary[side] = {
      counts,
      accuracy: gameAccuracy(sideMoves.map((m) => m.accuracy), wins),
      avg_cpl: sideMoves.length
        ? Math.round(sideMoves.reduce((a, m) => a + m.cpl, 0) / sideMoves.length)
        : null,
      avg_cpl_by_phase: avgByPhase(sideMoves),
    };
  }

  const keyMoments = moves
    .filter((m) => ["blunder", "mistake", "miss"].includes(m.class))
    .sort((a, b) => b.win_before - b.win_after - (a.win_before - a.win_after))
    .slice(0, 6);

  const h = game.getHeaders();
  return {
    white: h.White || "Brancas",
    black: h.Black || "Pretas",
    white_elo: h.WhiteElo,
    black_elo: h.BlackElo,
    result: h.Result || "*",
    opening: openingName(h),
    date: h.UTCDate || h.Date || "",
    my_color: myColor,
    depth,
    moves,
    summary,
    key_moments: keyMoments,
    eval_series: [
      { ply: 0, eval: moves.length ? moves[0].eval_before : 0 },
      ...moves.map((m) => ({ ply: m.ply, eval: m.eval_after })),
    ],
  };
}

function avgByPhase(sideMoves) {
  const out = {};
  for (const phase of ["opening", "middlegame", "endgame"]) {
    const vals = sideMoves.filter((m) => m.phase === phase).map((m) => m.cpl);
    if (vals.length) out[phase] = Math.round(vals.reduce((a, b) => a + b, 0) / vals.length);
  }
  return out;
}
